### PR TITLE
optional "failure_onion" in reply to htlc_accepted hook.

### PR DESCRIPTION
**Pull request by fiatjaf**
I need this feature to be able to use "shadow" last hops with fake nodes on [@lntxbot](https://t.me/lntxbot) and [Etleneum](https://etleneum.com) and return proper `incorrect_or_unknown_payment_details` for the pre-pay probes with the wrong `payment_hash` that Zap, Strike and BoS wallets send. These wallets will not attempt the actual payment if the pre-pay probe fail with any other error -- and it must come from the final destination, which means that in my case I must parse `onion.next_onion` and extract the ephemeral key, compute the shared secret then encode the onion and return it in the special field `"failure_onion"` this PR introduces.

The above paragraph is just for the context.

In reality a feature like this is also necessary if we are to write plugins that interface between two different Lightning Networks or similar networks, since a bridge node must be able to forward a failure onion from one network to the other and this PR (hopefully, maybe) enables that kind of thing.

Disclaimer: I have no idea of what I'm doing, but it works and is live on [@lntxbot](https://t.me/lntxbot). So maybe someone who knows what they're doing can think of a better way to implement this.

### CODE REVIEW

**Comment by fiatjaf** [comment]
I don't know how to write a test for this in Python. I already spent 4 days trying to make my Go code work and I made some heavy usage of lnd's libraries. Here's the part I parse the onion and build the error reply, if anyone is interested: https://github.com/fiatjaf/lntxbot/blob/bdfd13257334ec21411280ad06c9847f7ca4b0bf/htlc_accepted.go#L79-L137

**Comment by jsarenik** [comment]
Wow. This is interesting! And a nice small patch. @m-schmoock have a look please.

**Comment by rustyrussell** [comment]
This is great!! All it needs is:

1. doc/PLUGINS.md so people know it exists.
2. Changelog-Added: htlc_accepted hook can now return custom failure_onion.

File: lightningd/peer_htlcs.c [hunk 1]
```diff
@@ -940,6 +940,16 @@ static bool htlc_accepted_hook_deserialize(struct htlc_accepted_hook_payload *re
 				      " hook: %.*s",
 				      failmsgtok->end - failmsgtok->start,
 				      buffer + failmsgtok->start);
+        } else if ((failoniontok = json_get_member(buffer, toks,
```
Comment (m-schmoock) [inline lightningd/peer_htlcs.c:943]: Nit: indentations are tabs in our project. Note: We use a visual tab-width of 8 spaces.

File: doc/PLUGINS.md [hunk 2]
```diff
@@ -1001,7 +1001,23 @@ onion fields which a plugin doesn't want lightningd to consider.
...
+
+
+```json
+{
+  "result": "fail",
+  "failure_onion": "[292bytes of a serialized error packet]"
+}
+```
+
+If instead of `failure_message` the response contain a hex-encoded
```
Comment (m-schmoock) [inline doc/PLUGINS.md:1015]:
```suggestion
Instead of `failure_message` the response can contain a ...
```

**Review by m-schmoock — CHANGES_REQUESTED** [review]
Great!
Just some nits and I would like to have a pytest blackbox test for it in e.g. `tests/test_plugin.py` where we do all of the plugin API checks.
If you need assistance ping me.

File: doc/PLUGINS.md [hunk 3]
```diff
@@ -1001,7 +1001,23 @@ onion fields which a plugin doesn't want lightningd to consider.
...
+`failure_onion` that will be used instead (please refer to the
+[spec][bolt4-failure-onion] for details). Note that `lightningd` will
+apply the obfuscation step to the value retrned here with its own shared
+secret (and key type `ammag`) before returning it to the previous hop.
+
```
Comment (m-schmoock) [inline doc/PLUGINS.md:1020]: Maybe you could also point out the 'LN bridging nodes' usecase for this here.

File: lightningd/peer_htlcs.c (hunk 1 above)
Comment (fiatjaf) [reply lightningd/peer_htlcs.c:943]: In which circumstances are spaces used? I do see them in many lines mixed with the tabs.

Comment (m-schmoock) [reply lightningd/peer_htlcs.c]: Ah, you can refer to `doc/STYLE.md` (Tabs and indentation=

Comment (m-schmoock) [reply lightningd/peer_htlcs.c]: Basically its tabs with a visual length of 8 spaces. If you break up e.g. a function call to align the arguments each line then you can use spaces (after the tabs) to make them start at exactly the indentation you need.

Comment (fiatjaf) [reply lightningd/peer_htlcs.c]: Thank you.
These are rules very hard to follow, but I like them.

File: doc/PLUGINS.md [hunk 4]
```diff
@@ -1001,7 +1001,25 @@ onion fields which a plugin doesn't want lightningd to consider.
...
+  "result": "fail",
+  "failure_onion": "[292bytes of a serialized error packet]"
+}
+```
+
+Instead of `failure_message` the response can contain a hex-encoded
+`failure_onion` that will be used instead (please refer to the
+[spec][bolt4-failure-onion] for details). This can be used, for example,
+if you're writing a bridge between two Lightning Networks. Note that
+`lightningd` will apply the obfuscation step to the value retrned here
```
Comment (cdecker) [inline doc/PLUGINS.md:1019]:
```suggestion
`lightningd` will apply the obfuscation step to the value returned here
```

**Comment by fiatjaf** [comment]
Tried to add a test and discovered that:

1. Returning an invalid onion causes `lightningd` to crash, which is more-or-less expected, but I don't know how to account for this in the tests.
2. There is a memory leak on `lightningd` upon received the `failure_onion`:

```
E           ValueError:
E           Node errors:
E           Global errors:
E            - Node /tmp/ltests-lpfkw05m/test_htlc_accepted_hook_failonion_1/lightning-2/
 has memory leaks: [
E               {
E                   "backtrace": [
E                       "ccan/ccan/tal/tal.c:442 (tal_alloc_)",
E                       "ccan/ccan/tal/tal.c:471 (tal_alloc_arr_)",
E                       "common/json.c:244 (json_tok_bin_from_hex)",
E                       "lightningd/peer_htlcs.c:946 (htlc_accepted_hook_deserialize)",
E                       "lightningd/plugin_hook.c:184 (plugin_hook_callback)",
E                       "lightningd/plugin.c:367 (plugin_response_handle)",
E                       "lightningd/plugin.c:473 (plugin_read_json_one)",
E                       "lightningd/plugin.c:518 (plugin_read_json)",
E                       "ccan/ccan/io/io.c:59 (next_plan)",
E                       "ccan/ccan/io/io.c:407 (do_plan)",
E                       "ccan/ccan/io/io.c:417 (io_ready)",
E                       "ccan/ccan/io/poll.c:445 (io_loop)",
E                       "lightningd/io_loop_with_timers.c:24 (io_loop_with_timers)",
E                       "lightningd/lightningd.c:1015 (main)"
E                   ],
E                   "label": "common/json.c:244:u8[]",
E                   "parents": [],
E                   "value": "0x5594791ec358"
E               }
E           ]
```

I therefore give up.

File: lightningd/peer_htlcs.c [hunk 5]
```diff
@@ -940,6 +940,16 @@ static bool htlc_accepted_hook_deserialize(struct htlc_accepted_hook_payload *re
...
+		} else if ((failoniontok = json_get_member(buffer, toks,
+							       "failure_onion"))) {
+			failonion = json_tok_bin_from_hex(NULL, buffer, failoniontok);
+			if (!failonion)
+				fatal("Bad failure_onion for htlc_accepted"
+				      " hook: %.*s",
+				      failoniontok->end -  failoniontok->start,
+				      buffer + failoniontok->start);
+			fail_in_htlc(hin, take(new_onionreply(tmpctx, failonion)));
+			return false;
```
Comment (cdecker) [inline lightningd/peer_htlcs.c:952]: The only difference between `failure_onion` and `failure_msg` is that the latter gets serialized with a prefix and HMAC'd. Would it make sense to pull `local_fail_in_htlc` into this function (basically just the `create_onionreply` call) and handle both cases the same way? That'd mean that we take an eventual `failure_msg`, serialize it using the `create_onionreply` and then store it in `failure_onion`. 

Also I think that `failure_onion` needs to take precedence over `failure_msg`, since more work went into building that one. Ideally we'd panic if both are set, but since this is a hook call, our options are limited, so logging an error is likely the best we can do here.

File: doc/PLUGINS.md [hunk 6]
```diff
@@ -1001,7 +1001,25 @@ onion fields which a plugin doesn't want lightningd to consider.
...
 `fail` will tell `lightningd` to fail the HTLC with a given hex-encoded
-`failure_message` (please refer to the [spec][bolt4-failure-messages] for details: `incorrect_or_unknown_payment_details` is the most common).
+`failure_message` (please refer to the [spec][bolt4-failure-messages] for
+details: `incorrect_or_unknown_payment_details` is the most common).
+
+
+```json
+{
+  "result": "fail",
+  "failure_onion": "[292bytes of a serialized error packet]"
```
Comment (cdecker) [inline doc/PLUGINS.md:1011]: The error onion has a variable size, so it'd be better not allude to a fixed size here by putting the `292bytes` here.

**Review by cdecker — COMMENTED** [review]
Very nice, I had missed that `failure_msg` is not serialized and lacking the HMAC when injecting a previous onion, so this is indeed needed.

Just some minor cleanups.

File: lightningd/peer_htlcs.c (hunk 1 above)
Comment (m-schmoock) [reply lightningd/peer_htlcs.c]: Check the `.editorconfig` in the repos basedir. My VIM editor respects it.

File: doc/PLUGINS.md (hunk 6 above)
Comment (fiatjaf) [reply doc/PLUGINS.md:1011]: Does it? I was pretty sure it had a fixed size.

This proves I shouldn't be writing these things.

Comment (m-schmoock) [reply doc/PLUGINS.md:1011]: You are for sure not more or less talented than I am :)
It's about learning not knowing

Comment (fiatjaf) [reply doc/PLUGINS.md:1011]: Why has it a variable size? The spec says it is 260 bytes plus an HMAC of 32 bytes, then it is XORed against a cipherstream of the same size on every hop. I don't understand how it can be different than 292 bytes.

File: lightningd/peer_htlcs.c (hunk 5 above)
Comment (fiatjaf) [reply lightningd/peer_htlcs.c]: What about this? https://github.com/fiatjaf/lightning/blob/9af91a8493be3eb3b41d422973f2faf6209a02d2/lightningd/peer_htlcs.c#L930-L988

It's not cool, but I think it is clear. And once the deprecated branch is removed it will be much better.

**Comment by niftynei** [comment]
fixed build error, addressed two outstanding review comments.

**Comment by cdecker** [comment]
ACK 51c3441d27a5cf5f1cdeebe0bf51b8780789a89a

File: lightningd/peer_htlcs.c [hunk 7]
```diff
@@ -929,9 +929,29 @@ static bool htlc_accepted_hook_deserialize(struct htlc_accepted_hook_payload *re
...
 	if (json_tok_streq(buffer, resulttok, "fail")) {
 		u8 *failmsg;
-		const jsmntok_t *failmsgtok, *failcodetok;
+		const jsmntok_t *failoniontok, *failmsgtok, *failcodetok;
 
+		failoniontok = json_get_member(buffer, toks, "failure_onion");
 		failmsgtok = json_get_member(buffer, toks, "failure_message");
+
+		if (failoniontok) {
+			failonion = json_tok_bin_from_hex(tmpctx, buffer,
```
Comment (niftynei) [inline lightningd/peer_htlcs.c:938]: @fiatjaf the memory leak was caused by this line

```original
			failonion = json_tok_bin_from_hex(NULL, buffer,
```

By allocating it to `NULL` this means you're planning to manually clean it up; instead we use the magic allocation context `tmpctx`, which gets cleaned up, via magic.

File: lightningd/peer_htlcs.c [hunk 8]
```diff
@@ -929,9 +929,29 @@ static bool htlc_accepted_hook_deserialize(struct htlc_accepted_hook_payload *re
...
+				      " hook: %.*s",
+				      failoniontok->end -  failoniontok->start,
+				      buffer + failoniontok->start);
+
+			if (failmsgtok)
+				log_broken(ld->log, "Both 'failure_onion' and"
+					   "'failure_message' provided."
+					   " Ignoring 'failure_message'.");
+
+			fail_in_htlc(hin, take(new_onionreply(NULL,
```
Comment (niftynei) [inline lightningd/peer_htlcs.c:951]: when you `take` an object, then is a good time to allocate on `NULL`, as the `take` will reparent it to something else.

File: doc/PLUGINS.md (hunk 6 above)
Comment (fiatjaf) [reply doc/PLUGINS.md]: Here: https://github.com/lightningnetwork/lightning-rfc/blob/master/04-onion-routing.md#returning-errors

It says the packet consists of `hmac(32) + failure_len(2) + failure + pad_len(2) + pad` such that `failure + pad = 256`. The `channel_update` and other things I imagine are contained in the `failure` element above.

File: lightningd/peer_htlcs.c (hunk 8 above)
Comment (fiatjaf) [reply lightningd/peer_htlcs.c:951]: I think I get it now, kinda.
Thank you.

**Comment by niftynei** [comment]
ACK 6aa61e3cc3edb35deeff583ce321fdcc119716e4

**Comment by m-schmoock** [comment]
Really awesome :D :+1:
